be/src/olap/cumulative_compaction_time_series_policy.cpp

```diff
@@ -0,0 +1,255 @@
+                break;
+            }
+
+            bool is_delete = rs->has_delete_predicate();
+
+            // break the loop if segments in this rowset is overlapping.
+            if (!is_delete && rs->is_segments_overlapping()) {
+                *ret_cumulative_point = rs->version().first;
+                break;
+            }
+
+            // check the rowset is whether less than _compaction_goal_size
+            // The result of compaction may be slightly smaller than the _compaction_goal_size.
+            if (!is_delete && rs->version().first != 0 &&
+                rs->total_disk_size() < (_compaction_goal_size * 0.8)) {
+                *ret_cumulative_point = rs->version().first;
+                break;
+            }
+
+            // include one situation: When the segment is not deleted, and is singleton delta, and is NONOVERLAPPING, ret_cumulative_point increase
+            prev_version = rs->version().second;
+            *ret_cumulative_point = prev_version + 1;
+        }
+        VLOG_NOTICE
+                << "cumulative compaction time serires policy, calculate cumulative point value = "
+                << *ret_cumulative_point << " tablet = " << tablet->full_name();
+    }
+}
+
+int TimeSeriesCumulativeCompactionPolicy::pick_input_rowsets(
+        Tablet* tablet, const std::vector<RowsetSharedPtr>& candidate_rowsets,
+        const int64_t max_compaction_score, const int64_t min_compaction_score,
+        std::vector<RowsetSharedPtr>* input_rowsets, Version* last_delete_version,
+        size_t* compaction_score) {
+    if (tablet->tablet_state() == TABLET_NOTREADY) {
+        return 0;
+    }
+
+    int transient_size = 0;
+    *compaction_score = 0;
+    input_rowsets->clear();
+    int64_t total_size = 0;
+
+    for (auto& rowset : candidate_rowsets) {
+        // check whether this rowset is delete version
+        if (rowset->rowset_meta()->has_delete_predicate()) {
+            *last_delete_version = rowset->version();
+            if (!input_rowsets->empty()) {
+                // we meet a delete version, and there were other versions before.
+                // we should compact those version before handling them over to base compaction
+                break;
+            } else {
+                // we meet a delete version, and no other versions before, skip it and continue
+                input_rowsets->clear();
+                *compaction_score = 0;
+                transient_size = 0;
+                continue;
+            }
+        }
+
+        *compaction_score += rowset->rowset_meta()->get_compaction_score();
+        total_size += rowset->rowset_meta()->total_disk_size();
+
+        transient_size += 1;
+        input_rowsets->push_back(rowset);
+
+        // Condition 1: the size of input files for compaction meets the requirement of parameter _compaction_goal_size
+        if (total_size >= _compaction_goal_size) {
+            return transient_size;
```

Comment:
Is it necessary to consider last_delete_version before return?

Reply:
last_delete_version has a separate branch